dev-tools/mage/manifest/manifest.go

```diff
@@ -251,76 +186,119 @@ func DownloadComponents(ctx context.Context, manifest string, platforms []string
+	relaxedPackageName := regexp.QuoteMeta(packageName[:versionIndex])
+	relaxedPackageName += relaxedVersion
+	relaxedPackageName += regexp.QuoteMeta(packageName[versionIndex+len(dependencyVersion):])
+
+	if mg.Verbose() {
+		log.Printf(">>>>>>>>>>> Attempting to match a filename with %s", relaxedPackageName)
+	}
+
+	relaxedPackageNameRegexp, err := regexp.Compile(relaxedPackageName)
+	if err != nil {
+		return nil, fmt.Errorf("compiling relaxed package name regex %q: %w", relaxedPackageName, err)
+	}
+
+	for pkgName, pkg := range project.Packages {
+		if mg.Verbose() {
+			log.Printf(">>>>>>>>>>> Evaluating filename %s", pkgName)
+		}
+		if relaxedPackageNameRegexp.MatchString(pkgName) {
+			if mg.Verbose() {
+				log.Printf(">>>>>>>>>>> Found matching packageName for [%s, %s]: %s", project.Branch, project.CommitHash, pkgName)
+			}
+			return &ResolvedPackage{
+				Name: pkgName,
+				URLs: []string{pkg.URL, pkg.ShaURL, pkg.AscURL},
+			}, nil
+		}
 	}
 
-	return []string{val.URL, val.ShaURL, val.AscURL}, nil
+	return nil, fmt.Errorf("package [%s] not found in project manifest at %s", packageName, project.ExternalArtifactsManifestURL)
+}
+
+// versionRegexp is taken from https://semver.org/ (see the FAQ section/Is there a suggested regular expression (RegEx) to check a SemVer string?)
+const versionRegexp = `^(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(0|[1-9]\d*)(?:-(?:(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?:[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$`
+const anyPatchVersionRegexp = `(?:0|[1-9]\d*)`
+
+var versionRegExp = regexp.MustCompile(versionRegexp)
+
+func relaxVersion(version string) (string, error) {
+	matchIndices := versionRegExp.FindSubmatchIndex([]byte(version))
+	// Matches index pairs are (0,1) for the whole regexp and (2,3) for the patch group
+	// check that we have matched correctly
+	if len(matchIndices) < 4 {
+		return "", fmt.Errorf("failed to match regexp for version [%s]", version)
+	}
+
+	// take the starting index of the patch version
+	patchStartIndex := matchIndices[2]
+	// copy everything before the patch version escaping the regexp
+	relaxedVersion := regexp.QuoteMeta(version[:patchStartIndex])
+	// add the patch regexp
+	relaxedVersion += anyPatchVersionRegexp
+	// check if there's more characters after the patch version
+	remainderIndex := matchIndices[3]
+	if remainderIndex < len(version) {
+		// This is a stricter regexp that allows flexibility ONLY on the patch version (prerelease and build metadata must be the same)
+		// if we have a remainder from the original version, add it escaping it once more
+		// relaxedVersion += regexp.QuoteMeta(version[remainderIndex:])
+
+		// This is a looser regexp that allows anything beyond the major version to change (while still enforcing a valid patch version though)
+		// TODO: check with @dwhyrock why we allow more than just the patch version to change
```

Comment:
Not fully sure what is meant by this.

For the IAR test version you have (`8.14.0+build202406201002`), most of the components ("opted-out") should have the same `Major.Minor.Patch` version as the IAR version (`8.14.0`).  

Any "opted-in" projects (`endpoint-dev`) should be exactly 1 patch version higher (`8.14.1`).  Nothing else should change.

It's also worth noting that none of the components in the IAR manifest will have the build metadata (`+build202406201002`).  That's only for the top level of the manifest.

Does that clear some things up?  Or have I misunderstood what you were saying here?

Reply:
At this time the loose implementation works in the Independent Agent Release scenario.
For the unified release, all dependencies have the same version so we always get an exact match.

I don't really have a preference and it's not gonna be a problem if we don't have to deal with multiple versions of the same package (today this is not allowed in IAR or unified release manifests)

I am going to remove the TODO and leave the loose implementation in place.
With the `relaxVersion` function in place, we can fine-tune the regexp generation to suit more advanced cases if we need to.

Reply:
Sounds great to me.